File: docs/solver.js
// Number Cracker solver with dual feedback.
// ------------------------------------------------------------------
// Game: your friend picks a 3- or 4-digit secret (repeats and leading
// zeros allowed). Every guess returns two numbers:
//   - inOrder: digits in the right position (classic "bulls")
//   - correct: total digits in the secret counting multiplicity,
//              including inOrder (classic bulls + cows)
// You win when inOrder === length.
//
// With repeats, each secret digit matches at most one guess digit —
// that's the standard scoring rule (no double-counting).

const OPENERS = { 3: "012", 4: "0123" };

const _candidateCache = new Map();
export function allCandidates(length) {
  if (!_candidateCache.has(length)) {
    const total = Math.pow(10, length);
    const list = [];
    for (let n = 0; n < total; n++) list.push(String(n).padStart(length, "0"));
    _candidateCache.set(length, Object.freeze(list));
  }
  return _candidateCache.get(length);
}

export function score(guess, secret) {
  let inOrder = 0;
  const guessCounts = new Array(10).fill(0);
  const secretCounts = new Array(10).fill(0);
  for (let i = 0; i < guess.length; i++) {
    if (guess[i] === secret[i]) {
      inOrder++;
    } else {
      guessCounts[Number(guess[i])]++;
      secretCounts[Number(secret[i])]++;
    }
  }
  let misplaced = 0;
  for (let d = 0; d < 10; d++) {
    misplaced += Math.min(guessCounts[d], secretCounts[d]);
  }
  return { correct: inOrder + misplaced, inOrder };
}

export function scoreKey(s) {
  return s.correct * 10 + s.inOrder;
}

export function isWin(s, length) {
  return s.inOrder === length;
}

export function isValidGuess(value, length) {
  return value.length === length && /^\d+$/.test(value);
}

export function filterCandidates(candidates, guess, observed) {
  const target = scoreKey(observed);
  return candidates.filter((c) => scoreKey(score(guess, c)) === target);
}

export class Solver {
  constructor(length = 3) {
    this.length = length;
    this.allGuesses = allCandidates(length);
    this.candidates = this.allGuesses;
    this.turn = 0;
    this.cache = new Map();
  }

  reset() {
    this.candidates = this.allGuesses;
    this.turn = 0;
  }

  setLength(length) {
    if (length === this.length) return;
    this.length = length;
    this.allGuesses = allCandidates(length);
    this.candidates = this.allGuesses;
    this.turn = 0;
    this.cache = new Map();
  }

  record(guess, observed) {
    this.candidates = filterCandidates(this.candidates, guess, observed);
    this.turn += 1;
  }

  nextGuess() {
    if (this.turn === 0) return OPENERS[this.length] || this.allGuesses[0];
    if (this.candidates.length === 0) return null;
    if (this.candidates.length <= 2) return this.candidates[0];
    const key = [...this.candidates].sort().join(",");
    const cached = this.cache.get(key);
    if (cached) return cached;
    const guess = this._minimaxGuess();
    this.cache.set(key, guess);
    return guess;
  }

  _minimaxGuess() {
    const candidateSet = new Set(this.candidates);
    let best = null;
    for (const guess of this.allGuesses) {
      const evaluation = this._evaluate(guess, candidateSet);
      if (best === null || this._isBetter(evaluation, best)) {
        best = evaluation;
      }
    }
    return best.guess;
  }

  _evaluate(guess, candidateSet) {
    const buckets = new Map();
    for (const secret of this.candidates) {
      const k = scoreKey(score(guess, secret));
      buckets.set(k, (buckets.get(k) || 0) + 1);
    }
    const total = this.candidates.length;
    let worstCase = 0;
    let entropy = 0;
    for (const count of buckets.values()) {
      if (count > worstCase) worstCase = count;
      const p = count / total;
      entropy -= p * Math.log2(p);
    }
    return {
      guess,
      worstCase,
      entropy,
      isCandidate: candidateSet.has(guess),
    };
  }

  _isBetter(a, b) {
    if (a.worstCase !== b.worstCase) return a.worstCase < b.worstCase;
    if (Math.abs(a.entropy - b.entropy) > 1e-9) return a.entropy > b.entropy;
    if (a.isCandidate !== b.isCandidate) return a.isCandidate;
    return a.guess < b.guess;
  }

  topGuesses(limit = 5) {
    const candidateSet = new Set(this.candidates);
    const evaluations = this.allGuesses.map((g) =>
      this._evaluate(g, candidateSet)
    );
    evaluations.sort((a, b) => {
      if (a.worstCase !== b.worstCase) return a.worstCase - b.worstCase;
      if (Math.abs(a.entropy - b.entropy) > 1e-9) return b.entropy - a.entropy;
      if (a.isCandidate !== b.isCandidate) return a.isCandidate ? -1 : 1;
      return a.guess < b.guess ? -1 : 1;
    });
    return evaluations.slice(0, limit);
  }
}

File: solver.py
"""Knuth-style minimax solver for Bulls and Cows."""

from __future__ import annotations

import math
from collections import Counter
from dataclasses import dataclass, field

from game import Score, all_candidates, filter_candidates, score


OPENING_GUESS = "0123"


@dataclass
class GuessEvaluation:
    guess: str
    worst_case: int
    expected_remaining: float
    entropy: float
    is_candidate: bool


@dataclass
class Solver:
    all_guesses: tuple[str, ...] = field(default_factory=all_candidates)
    candidates: tuple[str, ...] = field(default_factory=all_candidates)
    turn: int = 0
    _guess_cache: dict[frozenset[str], str] = field(default_factory=dict)

    def reset(self) -> None:
        self.candidates = self.all_guesses
        self.turn = 0

    def record(self, guess: str, observed: Score) -> None:
        """Shrink the candidate pool given a guess and its observed score."""
        self.candidates = filter_candidates(self.candidates, guess, observed)
        self.turn += 1

    def next_guess(self) -> str:
        """Pick the next guess using Knuth minimax with entropy tiebreak."""
        if self.turn == 0:
            return OPENING_GUESS
        if len(self.candidates) <= 2:
            return self.candidates[0]
        key = frozenset(self.candidates)
        cached = self._guess_cache.get(key)
        if cached is not None:
            return cached
        guess = self._minimax_guess()
        self._guess_cache[key] = guess
        return guess

    def _minimax_guess(self) -> str:
        candidate_set = set(self.candidates)
        best: GuessEvaluation | None = None
        for guess in self.all_guesses:
            evaluation = self._evaluate(guess, candidate_set)
            if best is None or self._is_better(evaluation, best):
                best = evaluation
        assert best is not None
        return best.guess

    def _evaluate(self, guess: str, candidate_set: set[str]) -> GuessEvaluation:
        buckets: Counter[Score] = Counter()
        for secret in self.candidates:
            buckets[score(guess, secret)] += 1
        total = len(self.candidates)
        worst_case = max(buckets.values())
        expected = sum(count * count for count in buckets.values()) / total
        entropy = -sum(
            (count / total) * math.log2(count / total) for count in buckets.values()
        )
        return GuessEvaluation(
            guess=guess,
            worst_case=worst_case,
            expected_remaining=expected,
            entropy=entropy,
            is_candidate=guess in candidate_set,
        )

    @staticmethod
    def _is_better(a: GuessEvaluation, b: GuessEvaluation) -> bool:
        """Minimax primary, then entropy, then prefer a guess that could win now."""
        if a.worst_case != b.worst_case:
            return a.worst_case < b.worst_case
        if abs(a.entropy - b.entropy) > 1e-9:
            return a.entropy > b.entropy
        if a.is_candidate != b.is_candidate:
            return a.is_candidate
        return a.guess < b.guess

    def top_guesses(self, limit: int = 5) -> list[GuessEvaluation]:
        """Return the top-N evaluated guesses — used for UI insight."""
        candidate_set = set(self.candidates)
        evaluations = [self._evaluate(g, candidate_set) for g in self.all_guesses]
        evaluations.sort(
            key=lambda e: (e.worst_case, -e.entropy, not e.is_candidate, e.guess)
        )
        return evaluations[:limit]
